lib/protocol2.js

```diff
@@ -0,0 +1,254 @@
+    this.closed = false
+
+    this._byDiscoveryKey = new Map()
+    this._byId = []
+    this._byRemoteId = []
+    this._opens = []
+
+    this.protocol = this.muxer.addProtocol({
+      context: this,
+      name: 'hypercore',
+      messages: 13,
+      onremoteclose: () => {
+        this.closed = true
+
+        const peers = this._byId
+
+        this._byId = []
+        this._byRemoteId = []
+        this._byDiscoveryKey.clear()
+
+        while (this._opens.length) {
+          this._opens.pop().done()
+        }
+
+        for (let i = 0; i < peers.length; i++) {
+          const peer = peers[i]
+          if (!peer) continue
+          this._callPeerClose(peer)
+        }
+      }
+    })
+
+    this.open = this.protocol.addMessage({ type: 0, encoding: m.wire.open, onmessage: onopen })
+    this.close = this.protocol.addMessage({ type: 1, encoding: m.wire.close, onmessage: onclose })
+    this.sync = this.protocol.addMessage({ type: 2, encoding: m.wire.sync, onmessage: onsync })
+    this.request = this.protocol.addMessage({ type: 3, encoding: m.wire.request, onmessage: onrequest })
+    this.cancel = this.protocol.addMessage({ type: 4, encoding: m.wire.cancel, onmessage: oncancel })
+    this.data = this.protocol.addMessage({ type: 5, encoding: m.wire.data, onmessage: ondata })
+    this.noData = this.protocol.addMessage({ type: 6, encoding: m.wire.noData, onmessage: onnodata })
+    this.want = this.protocol.addMessage({ type: 7, encoding: m.wire.want, onmessage: onwant })
+    this.unwant = this.protocol.addMessage({ type: 8, encoding: m.wire.unwant, onmessage: onunwant })
+    this.bitfield = this.protocol.addMessage({ type: 9, encoding: m.wire.bitfield, onmessage: onbitfield })
+    this.range = this.protocol.addMessage({ type: 10, encoding: m.wire.range, onmessage: onrange })
+    this.reorgHint = this.protocol.addMessage({ type: 11, encoding: m.wire.reorgHint, onmessage: onreorghint })
+  }
+
+  cork () {
+    this.protocol.cork()
+  }
+
+  uncork () {
+    this.protocol.uncork()
+  }
+
+  async _callPeerClose (peer) {
+    try {
+      await peer.onremoteclose()
+    } catch (err) {
+      safetyCatch(err)
+    }
+  }
+
+  _sync (id) {
+    return id < this._byRemoteId.length ? this._byRemoteId[id] : null
+  }
+
+  async _async (id) {
+    for (const p of this._opens) {
+      if (p.id === id) {
+        await p.promise
+        break
+      }
+    }
+    return this.closed ? null : this._sync(id)
+  }
+
+  _alloc () {
+    const id = this._byId.indexOf(null)
+    return id === -1 ? this._byId.push(null) - 1 : id
+  }
+
+  _setRemoteId (id, peer) {
+    // TODO: Some better abstraction for this that is safer, allow MAX backlog...
+    while (id >= this._byRemoteId.length) this._byRemoteId.push(null)
+    this._byRemoteId[id] = peer
+  }
+
+  _get (dk, upsert) {
+    const k = dk.toString('hex')
+
+    let map = this._byDiscoveryKey.get(k)
+    if (map || !upsert) return map
+
+    map = [0, 0]
```

Comment:
Could use a comment about what these values are

Reply:
localId / remoteId and yea

Reply:
fixed by moving this into protomux